### PR TITLE
Fix metered sync warning being shown if the user has no internet

### LINKED ISSUE #12258: Metered sync warning shows even if the user has no internet
The "You have no internet collection - Sync anyway" snackbar should be shown instead

### PR DESCRIPTION
**Pull request by BrayanDSO**
## Purpose / Description
The "You have no internet collection - Sync anyway" snackbar should be shown instead

## Fixes
Fixes #12258

## Approach
Add a method to check if the user is connected to the internet

## How Has This Been Tested?

https://user-images.githubusercontent.com/69634269/187761689-f59bedb8-2b0c-4ddb-a418-3ebd1de3dbd1.mp4

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt [hunk 1]
```diff
@@ -16,14 +16,42 @@
...
-        .applicationContext
-        .getSystemService<ConnectivityManager>()
-        ?.isActiveNetworkMetered
-        ?: true
+object NetworkUtils {
+    private val connectivityManager: ConnectivityManager? =
+        AnkiDroidApp
+            .instance
+            .applicationContext
+            .getSystemService()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt:29]: This should be a `get()` property - `AnkiDroidApp.instance` may be null, and we shouldn't access this in a static constructor

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt [hunk 2]
```diff
@@ -16,14 +16,42 @@
...
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            connectivityManager?.getNetworkCapabilities(connectivityManager.activeNetwork)
+                ?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+                ?: false
+        } else {
+            connectivityManager?.activeNetworkInfo
+                ?.isConnected
+                ?: false
+        }
+    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt:56]: Can this be factored into/with 

https://github.com/ankidroid/Anki-Android/blob/fab95690f27284a19ebcd5199b3f96aeee9b12a6/AnkiDroid/src/main/java/com/ichi2/async/Connection.kt#L578-L601

**Review by david-allison — CHANGES_REQUESTED** [review]
One change, one question. Actual fix seems solid

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt (hunk 2 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt:56]:
```kotlin
if (allowLoginSyncOnNoConnection) { 
             return true 
         } 
```
Probably better to keep the preference separated from the util. 

The rest looks good to me. Nice touch checking for suspension and validation. I'll do the changes

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt (hunk 1 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt:29]: As NetworkUtils is a `object`, its properties are only initialized when the object is first used (https://kotlinlang.org/docs/object-declarations.html#object-declarations-overview). Currently the two methods of NetworkUtils use it, so shouldn't be a problem. But I'll do the change to make this future-proof

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt (hunk 2 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt]: Okay, tried something like the proposed code. PTAL.

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt [hunk 3]
```diff
@@ -16,14 +16,51 @@
...
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                    isOnline && networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_SUSPENDED)
+                } else {
+                    isOnline
+                }
+            } else {
+                cm.activeNetworkInfo?.isConnected
+                    ?: false
+            }
+        }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt:65]: Is it feasible to have `Connection`'s code call this extracted method?

Having two pieces of similar code leads to changes being applied to one and not the other

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt (hunk 3 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt:65]: Oh, that's where that code came from. Noticed only now.

Joined both methods on 0b13be2c5f2ed7b40e77e5db21af0a17adb1f9e2. 

Haven't tested but logically it seems right to me, as `isOnline` was only called at `launchConnectionTask`, so the `allowLoginSyncOnNoConnection` check can be made there

**Review by david-allison — APPROVED** [review]

**Review by mikehardy — APPROVED** [review]
LGTM

**Comment by david-allison** [comment]
This one broke CI: https://github.com/ankidroid/Anki-Android/commit/185e6aa792e37bb1903644cc9a36f7eb857cc9d9

**Pull request merged**